### ISSUE TITLE
Behavior Regarding "yield next" is Dependent on When Mounting is Done

### ISSUE THREAD

**Issue by trisys3** [koajs/mount#30]
It seems that `koa-mount` has different behavior for statements declared after `yield next` in nested contexts depending on when each mount is done. For example,

    const koa = require('koa');
    const mount = require(`koa-mount`);

    const topApp = koa();
    const middleApp = koa();
    const bottomApp = koa();

    /* If we uncomment these, in either order, only "Bottom app before..." is called. Huh? */
    // topApp.use(mount(middleApp));
    // middleApp.use(mount(bottomApp));

    topApp.use(function *(next) {
        console.log('Top app before next");

        yield next;

        console.log('Top app after next');
    });

    /* If we uncomment these, in either order, only "Top app before...", "bottom app before..."
       and "Top app after..." get called. */
    // topApp.use(mount(middleApp));
    // middleApp.use(mount(bottomApp));

    middleApp.use(function *(next) {
        console.log('Middle app before next');

        yield next;

        console.log('Middle app after next');
    });

    /* If we uncomment these, in either order, all 5 console statements get logged. */
    // topApp.use(mount(middleApp));
    // middleApp.use(mount(bottomApp));

    bottomApp.use(function *(next) {
        console.log('Bottom app');

        /* If we uncomment this, no matter where the "mount" statements are, all 5 console
           statements get logged. */
        // yield next;
    });

    // If we uncomment these, in either order, all 5 console statements get logged. */
    // topApp.use(mount(middleApp));
    // middleApp.use(mount(bottomApp));

    topApp.listen(3000);

My problem with this behavior is that it is unexpected and inconsistent. If it were documented, I might have less of a problem with it, but it would still be inconsistent. The behavior I would expect is that, no matter where the "mount" statements are, "Top app before..."; "Middle app before..."; "Bottom app before..."; "Middle app after..."; "Top app after..."; would get logged, in that order; in other words, the behavior that happens when the "mount" statements are after `middleApp`'s middleware gets defined.

If all middlewares have a `yield next` statement, all console statements get logged, no matter where the "mount" statements are, so this is not an issue. Ideally, to me, the behavior regarding middleware stacks where any middleware does not have a `yield next` would be:

    1. All statements after `yield next` in middlewares in the same and higher contexts declared
        before the middleware without `yield next` should be executed.
    2. In middlewares after the middleware without `yield next`, all statements after `yield
        next` should not be executed if the middleware is in the same or lower contexts.
    3. For middlewares in contexts above the middleware without `yield next`, all statements
        should always run, no matter whether the middleware is declared before or after the
        middleware without `yield next`.

For middleware stacks with multiple middlewares without `yield next`, at different context levels, the behavior should still follow basically the same logic.

Please tell me if any of this is confusing or not right. Also please keep in mind this is only my opinion on an ideal workflow, and although I believe the current flow is not quite right, my suggested flow may not be the best for this middleware. I did not help write the code, I am just a concerned programming citizen who uses this middleware and happens to have noticed this issue.

**Comment by jonathanong** [comment]
this is quite dense to read... mind making a PR?

**Comment by trisys3** [comment]
Sorry, swamped at work. I'm not using `koa` at work, though I'm evaluating its uses for our site, but I am using it at home for a personal project. I tried to use `koa-mount` when I encountered this bug, then I simplified my site a lot, removing the need for `koa-mount`, then you responded, then I got a huge deadline, and I was not able to work on my home site as much. I might be mixing up events, it was 6 months ago. Either way, I'm sorry I haven't responded.

I'm thinking I was not using `koa-mount` in the way it was intended: I was sort of trying to make middleware out of a whole other `koa` application, which does not seem to be supported by `koa`. I will try to get a minimal version of what I was trying to do, but I suspect this will prevent it from being considered, at least until `koa` supports that use case.

What I was trying to do was have different "zones" for my site, each as a different `koa` application, and `koa-mount` sort of seemed to support that use case better than `koa` itself, until I encountered this bug. As I said, I simplified the site right after, keeping everything in 1 "zone", so I didn't need `koa-mount` anymore.

I am closing this issue until I can get a minimal test, as I suspect my issue would go away if I had used `koa` how it was intended to be used.
